src/test/statistics/chunk_statistics/counting_quotient_filter_test.cpp

```diff
@@ -0,0 +1,134 @@
+    int_value_counts[1234] = 21;
+    int_value_counts[12345] = 8;
+    int_value_counts[123456] = 6;
+    int_segment = std::make_shared<ValueSegment<int>>();
+    for (auto value_and_count : int_value_counts) {
+      for (size_t i = 0; i < value_and_count.second; i++) {
+        int_segment->append(value_and_count.first);
+      }
+    }
+
+    string_cqf2 = std::make_shared<CountingQuotientFilter<std::string>>(4, 2);
+    string_cqf4 = std::make_shared<CountingQuotientFilter<std::string>>(4, 4);
+    string_cqf8 = std::make_shared<CountingQuotientFilter<std::string>>(4, 8);
+    string_cqf16 = std::make_shared<CountingQuotientFilter<std::string>>(4, 16);
+    string_cqf32 = std::make_shared<CountingQuotientFilter<std::string>>(4, 32);
+    string_cqf2->populate(string_segment);
+    string_cqf4->populate(string_segment);
+    string_cqf8->populate(string_segment);
+    string_cqf16->populate(string_segment);
+    string_cqf32->populate(string_segment);
+
+    int_cqf2 = std::make_shared<CountingQuotientFilter<int>>(4, 2);
+    int_cqf4 = std::make_shared<CountingQuotientFilter<int>>(4, 4);
+    int_cqf8 = std::make_shared<CountingQuotientFilter<int>>(4, 8);
+    int_cqf16 = std::make_shared<CountingQuotientFilter<int>>(4, 16);
+    int_cqf32 = std::make_shared<CountingQuotientFilter<int>>(4, 32);
+    int_cqf2->populate(int_segment);
+    int_cqf4->populate(int_segment);
+    int_cqf8->populate(int_segment);
+    int_cqf16->populate(int_segment);
+    int_cqf32->populate(int_segment);
+  }
+
+  std::shared_ptr<CountingQuotientFilter<std::string>> string_cqf2;
+  std::shared_ptr<CountingQuotientFilter<std::string>> string_cqf4;
+  std::shared_ptr<CountingQuotientFilter<std::string>> string_cqf8;
+  std::shared_ptr<CountingQuotientFilter<std::string>> string_cqf16;
+  std::shared_ptr<CountingQuotientFilter<std::string>> string_cqf32;
+  std::shared_ptr<CountingQuotientFilter<int>> int_cqf2;
+  std::shared_ptr<CountingQuotientFilter<int>> int_cqf4;
+  std::shared_ptr<CountingQuotientFilter<int>> int_cqf8;
+  std::shared_ptr<CountingQuotientFilter<int>> int_cqf16;
+  std::shared_ptr<CountingQuotientFilter<int>> int_cqf32;
+  std::shared_ptr<ValueSegment<std::string>> string_segment;
+  std::shared_ptr<ValueSegment<int>> int_segment;
+  std::map<std::string, size_t> string_value_counts;
+  std::map<int, size_t> int_value_counts;
+
+  template <typename DataType>
+  void test_value_counts(std::shared_ptr<CountingQuotientFilter<DataType>> cqf, std::map<DataType, size_t>
+      value_counts) {
+    for (auto value_and_count : value_counts) {
+        EXPECT_TRUE(cqf->count(value_and_count.first) >= value_and_count.second);
+    }
+  }
+
+  template <typename DataType>
+  void test_can_prune(std::shared_ptr<CountingQuotientFilter<DataType>> cqf, std::map<DataType, size_t>
+      value_counts) {
+    for (auto value_and_count : value_counts) {
+        EXPECT_FALSE(cqf->can_prune(value_and_count.first, PredicateCondition::Equals));
+    }
```

Comment:
There should be a test for successful pruning.

Reply:
The problem is that pruning is probabilistic, so that the test would sometimes randomly fail.

Reply:
Wait, does that mean it is possible that a CQF decides to prune a predicate even though that leads to an incorrect query result?
I thought this error class was not possible.

Reply:
You're right, I thought about it the wrong way, sorry for the confusion.

Reply:
How about this:
say we have a CQF containing just two values, let's say 1 and 10.
Is it possible that `can_prune(v) == false` for all values in [2, 9]?
Maybe we cannot say for sure that `can_prune(2) == true`, but if we can check that at least one of them is always prunable, that could be a start.

Reply:
I just added some false positive rate tests. Can you take a look?